src/groth16/circuit/builder/types.rs

```diff
@@ -0,0 +1,333 @@
+/// you can think of as the first Word8 on the left.
+///
+/// FromIterator: bytes input into from right to left; least significant first to
+/// most significant byte.
+///
+pub type Word64 = [Word8; 8];
+
+/// Rotates a Word64's bits by moving bit a position `i` into position `i+by`
+/// modulo the lane size. The least significant bit is where i = 0 and the most
+/// significant bit is where i = 63.
+///
+/// Example of left rotation: (u64 truncated to u8)
+///     start:   0100 0101
+///     becomes: 1000 1010
+///
+/// NOTE: in memory this would look like [1,0,1,0,0,0,1,0] => [0,1,0,1,0,0,0,1]
+///
+pub fn rotate_word64_left(input: Word64, by: usize) -> Word64 {
+    let mut wrd64 = Word64::default();
+    input
+        .iter()
+        .flat_map(|x| x.iter())
+        .cycle()
+        .skip(64 - (by % 64))
+        .take(64)
+        .zip(iproduct!(0..8, 0..8))
+        .for_each(|(&wire_id, (i, j))| wrd64[i][j] = wire_id);
+    wrd64
+}
+
+/// Rotates a Word64's bits by moving bit a position `i` into position `i-by`
+/// modulo the lane size.
+///
+/// Example of left rotation: (u64 truncated to u8)
+///     start:   0100 0101
+///     becomes: 1010 0010
+///
+/// NOTE: in memory this would look like [1,0,1,0,0,0,1,0] => [0,1,0,0,0,1,0,1]
+///
+pub fn rotate_word64_right(input: Word64, by: usize) -> Word64 {
+    let mut wrd64 = Word64::default();
+    input
+        .iter()
+        .flat_map(|x| x.iter())
+        .cycle()
+        .skip(by % 64)
+        .take(64)
+        .zip(iproduct!(0..8, 0..8))
+        .for_each(|(&wire_id, (i, j))| wrd64[i][j] = wire_id);
+    wrd64
+}
+
+/// NOTE: if you don't give enough bits the extra bits will be filled in with
+/// WireId::default() which is the zero wire. On the other hand if you have too
+/// many bits you will get a runtime panic! (this is so you realize you have
+/// made a grave mistake). The only way to get the panic that you have
+/// given too many bits is to collect a word64 directly.
+///
+pub fn to_word64(input: impl Iterator<Item = WireId>) -> Word64 {
+    let mut arr: Word64 = Word64::default();
+    input
+        .chunks(8)
+        .into_iter()
+        .map(|chunk| to_word8(chunk))
+        .zip_longest(0..8)
+        .for_each(|x| match x {
+            Both(num, i) => arr[i] = num,
+            Right(i) => panic!("to_word64: Word64 cannot be constructed from less than 64 WireId"),
+            Left(_) => panic!("to_word64: Word64 cannot be constructed from more than 64 WireId"),
+        });
+    arr
+}
+
+// pub enum Binary {
+//     One,
+//     Zero,
+// }
+
+// impl Binary {
+//     pub fn from_u64(num: u64) -> [Binary; 64] {
+//         let mut n = num;
+//         let mut output = [Zero; 64];
+//         (0..64).for_each(|i| {
+//             if n % 2 != 0 {
+//                 output[i] = One;
+//             }
+//             n = n >> 1;
+//         });
+//         output
+//     }
+// }
+
+pub const RC: [u64; 24] = [
+    1u64,
```

Comment:
Purely a style comment, ignore if you disagree: the type of each of the numbers can probably be inferred from the fact that `RC: [u64; 24]` so you can drop the `u64` suffix. Additionally, it might be more "aesthetic" if every number was padded with zeros to be the same length.

Reply:
Scared changing constants, but with all the tests it should be fine.